protocols/noise/src/io.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::Poll;
+use log::{debug, trace};
+use snow;
+use static_assertions::const_assert_eq;
+use std::{fmt, io};
+use tokio_io::{AsyncRead, AsyncWrite};
+
+const MAX_NOISE_PKG_LEN: usize = 65535;
+const MAX_WRITE_BUF_LEN: usize = 16384;
+
+// Buffer indices ///////////////////////////////////////////////////////////
+
+// Read buffer: [0, 65535)
+const READ_BEGIN: usize = 0;
+const READ_END: usize = READ_BEGIN + MAX_NOISE_PKG_LEN;
+
+// Decryption buffer for data read: [65535, 131070)
+const READ_CRYPTO_BEGIN: usize = READ_END;
+const READ_CRYPTO_END: usize = READ_CRYPTO_BEGIN + MAX_NOISE_PKG_LEN;
+
+// Write buffer: [131070, 147454)
+const WRITE_BEGIN: usize = READ_CRYPTO_END;
+const WRITE_END: usize = WRITE_BEGIN + MAX_WRITE_BUF_LEN;
+
+// Encryption buffer for data to write: [147454, 180222)
+const WRITE_CRYPTO_BEGIN: usize = WRITE_END;
+const WRITE_CRYPTO_END: usize = WRITE_CRYPTO_BEGIN + 2 * MAX_WRITE_BUF_LEN;
+
+const TOTAL_BUFFER_LEN: usize = WRITE_CRYPTO_END;
+
+// Static invariants ////////////////////////////////////////////////////////
+
+const_assert_eq!(A; READ_END - READ_BEGIN, MAX_NOISE_PKG_LEN);
+const_assert_eq!(B; READ_CRYPTO_END - READ_CRYPTO_BEGIN, MAX_NOISE_PKG_LEN);
+const_assert_eq!(C; WRITE_END - WRITE_BEGIN, MAX_WRITE_BUF_LEN);
+const_assert_eq!(D; WRITE_CRYPTO_END - WRITE_CRYPTO_BEGIN, 2 * MAX_WRITE_BUF_LEN);
+const_assert_eq!(E; 180222, TOTAL_BUFFER_LEN);
+const_assert_eq!(F; 0, READ_CRYPTO_BEGIN - READ_END); // read crypto buffer follows read buffer
+const_assert_eq!(G; 0, WRITE_CRYPTO_BEGIN - WRITE_END); // write crypto buffer follows write buffer
+
+pub struct NoiseOutput<T> {
```

Comment:
I'm not a fan of calling `poll_read` and `poll_write` with empty slices. It's a bit too weird.
I suggest we keep `NoiseOutput` as the "final" type (the one that the user receives), but add a `Handshake` type that wraps around `NoiseOutput`, performs the handshake, and has a `into_transport_mode()` method that turns it into a `NoiseOutput`. This way you can also make `session` private.